### PR TITLE
fix runtime err remove self or brother when onExit

### PR DESCRIPTION
**Pull request by litefeel**

### CODE REVIEW

**Comment by litefeel** [comment]
[ci rebuild]

**Comment by minggo** [comment]
how to reproduce it?

**Comment by litefeel** [comment]
1. create layer.
2. add sprite1, sprite2 to layer.
3. setOnExitCallback for sprite1, and remove sprite1 in the callback.
4. add layer to scene.
5. delay
6. remove the layer.
7. crash

**Comment by minggo** [comment]
It is not safe to remove/add a child in loop.
onExit is good for release some memory.

In theory, it is better to make all function safe to add/remove a child, such as `Node::enumerateChildren`, `Node::onEnter` and `onExit`, but it will make codes to complex to manage.

**Comment by litefeel** [comment]
in this case.

i have an function popMsg(), 
the function to do:
1. show an msg or icon or sh. on running scene
2. delay 3s
3. remove self.

when change scren, it must was removed, becase old scene will be show after sometime.

so. I must be remove it on `onExit`.

so. i want an safe `onExit`.

**Comment by litefeel** [comment]
I think more people need to remove the node in `onExit`.

**Comment by minggo** [comment]
A child will be removed by its parent when a scene is replaced.
Why you need to remove it in `onExit()`?

**Comment by litefeel** [comment]
1. runing scene is scene1.
2. show msg node in scene1.
3. pushScene scene2
4. popScene
5. running scene is scene1.

so, msg node still in scene1.

**Comment by minggo** [comment]
You remove msg node in popScene's onExit()?

**Comment by litefeel** [comment]
In fact, i will remove msg node in pushScene's onExit.

**Comment by minggo** [comment]
Yep, you remove it in scene2's onExit?

**Comment by litefeel** [comment]
remove msgNode in msgNode's onExit when call pushScene.

**Comment by minggo** [comment]
As i said, in theory it is right that every function can be invoked in every situation. But it will make engine too complex and slow performance. Or, we should modify everywhere when looping _children, and the codes seems tricky.

**Comment by litefeel** [comment]
add or remove child in onExit or onEnter is simple. 
so i think will be allow.

**Comment by litefeel** [comment]
you see my code. it is so simple and ease.

**Comment by minggo** [comment]
It is problem if invoking `removeFromParent` in `onExit`, `onExit` or `enumerateChildren`.

**Comment by litefeel** [comment]
now. can not call removeFromParent in enumerateChildren.
but call removeFromParent is safe in onEnter or onExit.

**Comment by litefeel** [comment]
i see. in c++.  first invoke callback . my test code in lua.

why invoke order in onExit:
1. c++ callback
2, js callback
3. other code
4. lua callback

**Comment by minggo** [comment]
But how to resolve the problem: remove more than one node in onExit()?

**Comment by litefeel** [comment]
后面我又想了想，对于`onEnter` `onExit` `enumerateChildren`的作用。
1. `onEnter`: 初始化或者为启动做准备，允许添加自身/兄弟/孩子节点，允许删除孩子节点，不允许删除自身/兄弟节点
2. `onExit`: 释放资源，允许删除自身/兄弟/孩子节点，不允许添加自身/兄弟/孩子节点
3. ·enumerateChildren`: 遍历节点，不允许添加删除自身/兄弟/孩子节点。

那么基于上面的思想。
onEnter函数体内的执行顺序为：
1. c++ js lua 的 callback
2. 其他代码

onExit函数体内的执行顺序为：
1. 其他代码
2. c++ js lua 的 callback

以上还有一个假设，一个Node仅有一个callback(c++ js lua)，或者最多仅有一个修改了节点树结构。

这次的代码还没有修改onEnter onExit的执行顺序，如果你同意我的看法，那么我就修改下。

**Comment by litefeel** [comment]
| op | `onEnter` | `onExit` | `enumerateChildren` |
| --- | --- | --- | --- |
| addChild | [o] | [o] | [o] |
| removeChild | [o] | [o] | [o] |
| addBrother | [o] | [x] | [x] |
| removeBrother | [x] | [o] | [x] |
| removeSelf | [x] | [o] | [x] |
- [x]: can not allow
- [o]: allow

now. i make the table.

**Comment by minggo** [comment]
Please paste english version by google translate if you are not confident with you English. Then all developers can discuss with it.
Thanks.

It sounds a good idea to do it like this. But it is very hard to remember in which situation can do what.

**Comment by minggo** [comment]
If we don't have a good solution about how to fix it more elegant, then let's keep as it is. Because it is simple, can not do remove/add operation in these functions.

**Comment by litefeel** [comment]
please see my table. It can achieve.

**Comment by minggo** [comment]
Yep, i have seen it.
How can developers know these limitation?

**Comment by minggo** [comment]
It is hard to remember, so many things should remember.

**Comment by litefeel** [comment]
- `onEnter`: do init sth, can add node to tree.
- `onExi`t: clear and free resource, can remove node to tree
- `enumerateChildren`:  for loop, can not change node tree.

by this mind.  this rules is so simple and ease.

**Comment by minggo** [comment]
But you table says addChild and removeChild can be used in enumerateChildren/onEnter.

**Comment by litefeel** [comment]
yes. it is allow by code that addChild/removeChild in enumerateChildren, but this rule is complex.
so remove this rule.

**Comment by minggo** [comment]
Got it.
So what's the final table?
Could you share it?

**Comment by litefeel** [comment]
| op | `onEnter` | `onExit` | `enumerateChildren` |
| --- | --- | --- | --- |
| addChild | [o] | [o] | [o] |
| removeChild | [o] | [o] | [o] |
| addBrother | [o] | [x] | [x] |
| removeBrother | [x] | [o] | [x] |
| removeSelf | [x] | [o] | [x] |
- [x]: can not allow
- [o]: allow

if have not this table. Use the following text description , it is simple but not accurate.
- `onEnter`: do init sth, can add node to tree.
- `onExit`: clear and free resource, can remove node from tree
- `enumerateChildren`:  for loop, can not change node tree.

**Comment by minggo** [comment]
The table is the same as before. It can invoke `addChild` and `removeChild` in `enumerateChildren` in the table.

**Comment by litefeel** [comment]
yes.

**Comment by minggo** [comment]
But you said

> enumerateChildren: for loop, can not change node tree.

They are conflict.

**Comment by litefeel** [comment]
我的意思是，这张表里的是最精确的，但是如果没有这张表，用简单的文字来简单描述的话，就不要在`enumerateChildren`中修改节点树，这是为了方便记忆，虽然这个文字描述是不精确的。

又或者，作为一个写程序的思想， `enumerateChildren`是为了遍历，查找，而不是为了修改节点树，
虽然代码上运行在`enumerateChildren`中添加子节点。

File: cocos/2d/CCNode.cpp [hunk 1]
```diff
@@ -1364,9 +1367,19 @@ void Node::onEnter()
...
+
+    _childrenDirty = false;
+    for (ssize_t i = 0; i < _children.size(); i++)
+    {
+        auto child = _children.at(i);
+        if (!child->isRunning())
+            child->onEnter();
+        if (_childrenDirty)
+        {
+            i = -1;
```
Comment (minggo) [inline cocos/2d/CCNode.cpp:1379]: It will loop from begin again?
Some children's `onEnter` will be invoked more than once.

Comment (litefeel) [inline cocos/2d/CCNode.cpp:1379]: if invoke `onEnter`, the child is running. so, it invoked `onEnter` once.

**Comment by minggo** [comment]
@lite3 
After more thinking about it i think i can not fix this PR, because
- it will slow down the performance
- it is not a right way to remove self or bother in `onExit()`

**Comment by litefeel** [comment]
this pr is not finish.

**Pull request closed without merging**